libcontainer/nsenter/nsexec.c

```diff
@@ -319,7 +320,74 @@ static int clone_parent(jmp_buf *env, int jmpval)
 		.jmpval = jmpval,
 	};
 
-	return clone(child_func, ca.stack_ptr, CLONE_PARENT | SIGCHLD, &ca);
+	/*
+	 * Since glibc 2.25 (see c579f48edba88380635ab98cb612030e3ed8691e),
+	 * glibc no longer updates the TLS state containing the current process
+	 * tid after clone(2). This results in stale TIDs being used when Go
+	 * 1.22 and later call pthread_gettattr_np(pthread_self()), resulting
+	 * in crashes on ancient glibcs and errors on newer glibcs.
+	 *
+	 * Luckily, because the same address is used for CLONE_PARENT_SETTID,
+	 * we can poke around in glibc's internal cache by getting the address
+	 * using PR_GET_TID_ADDRESS (only available in Linux >= 3.5, with
+	 * CONFIG_CHECKPOINT_RESTORE=y) and then overwriting it with
+	 * CLONE_CHILD_SETTID. CLONE_CHILD_CLEARTID is set to allow descendant
+	 * PR_GET_TID_ADDRESS calls to work, as well as matching what glibc
+	 * does in arch_fork().
+	 *
+	 * Yes, this is pretty horrific, but the core issue here is that we
+	 * need to run Go code ("runc init") in the child after fork(), which
+	 * is not allowed by glibc (see signal-safety(7)). We cannot exec to
+	 * solve the problem because we are in a security critical situation
+	 * here, and doing an exec would allow for container escapes (obvious
+	 * issues include that the shared libraries loaded from a re-exec would
+	 * come from the container, and doing an exec here would clear the bit
+	 * that makes non-dumpable flags effective for userns containers with
+	 * CAP_SYS_PTRACE).
+	 */
+
+	pid_t *tid_addr = NULL;
+	pid_t actual_tid = gettid();
+	if (prctl(PR_GET_TID_ADDRESS, &tid_addr) < 0) {
+		/*
+		 * We couldn't get &PTHREAD_SELF->tid, probably meaning we are running
+		 * with CONFIG_CHECKPOINT_RESTORE=n. Unfortunately the layout of
+		 * "struct pthread" is not public, but we need to get the address by
+		 * force.
+		 *
+		 * So, scan the structure as though it were pid_t[] to find the first
+		 * element which matches the actual tid of the current process. Yes,
+		 * this is *much* worse than PR_GET_TID_ADDRESS, but we should never
+		 * get here on the vast majority of machines.
+		 *
+		 * (To be honest, maybe it's better to just hope Go doesn't notice any
+		 * issues with glibc rather than trying to hack internal glibc
+		 * structures to make them "work" with Go. But it seems we need to do
+		 * this...)
+		 */
+
+		write_log(WARNING, "clone: PR_GET_TID_ADDRESS failed (%m): falling back to scanning pthread_self -- please use a kernel with CONFIG_CHECKPOINT_RESTORE=y");
+
+		pid_t *fake_tid_array = (pid_t *) pthread_self();
+		/* On my machine, fake_tid_array[180] is PTHREAD_SELF->tid. */
+		for (size_t i = 0; i < 512; i++) {
```

Comment:
What a wonderful solution!
I think this looks like `Credential-Stuffing attack`? Is there a chance that some other var's value happens equal to `actual_tid`?

Reply:
I'm not sure I'd call it "wonderful" :sweat_smile:

It is possible for another field to have the same value, and bad things might happen if we overwrite it. I've to implemented @kolyshkin's suggestion of trying some known field offsets first, and only doing a linear scan if the known offsets don't match (which should never happen in practice, unless glibc changes their structure layouts).

Reply:
> I'm not sure I'd call it "wonderful" 😅

I have tried `CLONE_CHILD_SETTID | CLONE_CHILD_CLEARTID`, but I didn't think out using it to rewrite the tid value in pthread_t. 👍